File: python/trees/trees_4.py
from trees_1 import TreeNode

# for tree node class definition see Minimal BST

# base case
# None => True

# recursive case
# node.data must fall within [left bound, right bound)
# left and right bounds are defined by the min / max ancestors
# left and right subtrees must satisfy isBST()


# O(n)
def isBST(tn, left_bound=None, right_bound=None):
    if not tn:
        return True

    # must be strictly greater than any ancestor to the left
    if left_bound is not None and tn.data <= left_bound:
        return False

    # must be less than or equal to any ancestor to the right
    if right_bound is not None and tn.data > right_bound:
        return False

    return isBST(tn.left, left_bound, tn.data) and isBST(tn.right, tn.data, right_bound)

File: go/pkg/trees/trees_4.go
package trees

// for TreeNode type def see Minimal BST

// IsBST runs in O(n)
// using pointers for the bounds to leverage
// a nil zero value rather than 0 which
// is valid data for a TreeNode<int>
func IsBST(tn *TreeNode, leftBound, rightBound *int) bool {
	if tn == nil {
		return true
	}

	if leftBound != nil && tn.Data <= *leftBound {
		return false
	}

	if rightBound != nil && tn.Data <= *rightBound {
		return false
	}

	return IsBST(tn.Left, leftBound, &tn.Data) && IsBST(tn.Right, &tn.Data, rightBound)
}
